sway-core/src/ir_generation/function.rs

```diff
@@ -57,6 +57,33 @@ use std::collections::HashMap;
 /// forced, into a value if that is desired. All the temporary values are manipulated with simple
 /// loads and stores, rather than anything more complicated like `mem_copy`s.
 
+// Wrapper around Value to enforce distinction between terminating and non-terminating values.
```

Comment:
I am rethinking this proposal now because "block temrinator" is also an established term. But still I am not comfortable with having "terminator" being any instruction at the and of the block and "is_terminator" in other cases. I don't know...

Reply:
> I am rethinking this proposal now because "block terminator" is also an established term. But still I am not comfortable with having "terminator" being any instruction at the and of the block and "is_terminator" in other cases. I don't know...

Yes, terminator is an established term for the last instruction in a block.

Reply:
Thanks Vaivas, yes, if it is already stablished we can keep it as it is, I am fine with that. Of course, as long as we are all aware that `get_terminator/_mut` can return a non-terminator.

The comment is already at the verge of bikeshedding, please forgive me if I am too much into this, but after looking at the usages of `get_terminator/_mut` I think we can still gain clarity if the `get_terminator` name stays the same, but the method returns `None` if the last instruction is not a terminator.

My reasoning behind that proposal:

1. All current usages of `get_terminator/_mut` inspect for a particular terminator and ignore the rest. A single usage of `get_terminator_mut` that doesn't is in `demote_block_signature` and expect already fully defined blocks to be well formed, i.e., to end in a terminator.<br>Means, if we do return `None` for non-terminators, we will not be breaking semantics of any of the existing callers and will be in sync with `is_terminator`.

1. LLVM [defines it the same way](https://llvm.org/doxygen/BasicBlock_8h_source.html):
```
/// Returns the terminator instruction if the block is well formed or null
/// if the block is not well formed.
const Instruction *getTerminator() const LLVM_READONLY {
    if (InstList.empty() || !InstList.back().isTerminator())
      return nullptr;
    return &InstList.back();
}
```

But as said, I am fine if we keep it as is. And please forgive me if I am bikeshedding to much here!

Reply:
>I think we can still gain clarity if the `get_terminator` name stays the same, but the method returns `None` if the last instruction is not a terminator.

This is very much reasonable. Note that, similar to how LLVM defines it, the IR is only valid if it returns `Some(_)`, and the verifier checks for it. It is only during IR construction that it can be invalid (i.e., return `None`).

Reply:
The conflation between "terminator" as "the final instruction of a block" and "terminator" as "specific instructions that are only allowed as block terminators" is already there in the code before this PR. The PR conflates the two concepts even more, but it's a problem that was already there to begin with.

I'm fine with the proposal. Do you want me to fix the `get_terminator` issue in this PR, or should I open up a separate one?

Reply:
Since it is a very simple change that do not affect the existing callers, I think it's fine to do it as a part of this PR.

Reply:
Will do.

Reply:
Fixed.